rfcs/0158-artifact-metadata.md

```diff
@@ -0,0 +1,62 @@
+# RFC 158 - artifact metadata
+* Comments: [#158](https://github.com/taskcluster/taskcluster-rfcs/pull/158)
+* Proposed by: @escapewindow
+
+# Summary
+
+The goal is to provide Artifact Integrity guarantees from the point the worker uploads an artifact, to the point that someone downloads that artifact to use. We can do this by:
+
+1. adding SHA metadata to artifacts in the Queue,
+2. ensuring that that metadata can't be modified once it's written, and
+3. providing a download tool that queries the Queue for an artifact's location and SHA, downloads the artifact, and verifies the downloaded SHA matches the SHA provided by the Queue.
+
+## Motivation
+
+This would improve robustness and security. By storing a SHA to verify on download, we can avoid corrupt downloads. By verifying the read-only SHA before we use the artifact, we can detect malicious tampering of artifacts at rest before we use it for a critical operation.
+
+(This would obsolete the [artifacts section of the Chain of Trust artifact](https://scriptworker.readthedocs.io/en/latest/chain_of_trust.html#chain-of-trust-artifacts), and improve artifact integrity platform-wide.)
```

Comment:
Is there a more detailed breakdown of how this provides all of the security features currently provided by CoT?

Reply:
CoT is mainly three things:

1. verifying that the artifacts which were uploaded are the same artifacts we're downloading,
2. verifying that upstream workers are under our control, and
3. verifying the task graph was generated from a trusted tree.

This RFC directly addresses the first point. We added the `artifacts` section of the CoT artifact to include the file path and sha256 of each artifact uploaded by each task. Because the CoT artifact itself wasn't covered, we added CoT artifact signatures, which kind of cover point 2 as well as help prove the CoT artifact itself wasn't modified at rest. But we only get a CoT artifact if we enable it in the task, and we only get a verifiable CoT signature on a limited set of workers. And currently only scriptworkers verify CoT.

By moving the artifact metadata into the platform, we have the capability of covering all task artifacts, not just the ones from level 3 trusted pools. And by moving artifact metadata verification out of scriptworker into a standalone download tool, we can use the same tool throughout the graph for full graph artifact integrity.

Reply:
(If worker [re]registration ends up being trusted enough, this RFC + worker [re]registration could end up obsoleting points 1 and 2 of CoT, and CoT can focus on verifying the graph came from a trusted tree.)

Reply:
Ok thank you for the detailed breakdown. I'd love to see this included in the RFC so that it can be referred back to later.

Reply:
Added, though I didn't go as far into detail about worker [re]registration, to keep the RFC more focused on artifact metadata.